File: scripts/test-districts.R
#!/usr/bin/env Rscript
# scripts/test-districts.R
# Pins the R district-id helpers to the shared R/JS fixtures
# (src/loaders/R/common/districts_fixtures.csv). The JS twin is
# scripts/test-districts.js — run BOTH after touching either implementation.

source("src/loaders/R/common/districts.R")

fx <- read.csv("src/loaders/R/common/districts_fixtures.csv",
               colClasses = "character", fileEncoding = "UTF-8")
failed <- 0L
for (i in seq_len(nrow(fx))) {
  got <- switch(fx$fn[i],
    maj = selfgov_from_maj_id(fx$input[i]),
    raw = selfgov_from_raw_district_id(fx$input[i]),
    dotted = council_maj_id_from_dotted(fx$input[i]),
    stop("unknown fixture fn: ", fx$fn[i])
  )
  if (as.character(got) != fx$expected[i]) {
    failed <- failed + 1L
    cat(sprintf("✗ %s(%s) = %s, expected %s  (%s)\n",
                fx$fn[i], fx$input[i], got, fx$expected[i], fx$note[i]))
  }
}
if (failed > 0L) {
  cat(sprintf("\n%d/%d fixture(s) FAILED\n", failed, nrow(fx)))
  quit(status = 1)
}
cat(sprintf("All %d district-id fixtures passed (R).\n", nrow(fx)))

File: scripts/test-districts.js
#!/usr/bin/env node
// scripts/test-districts.js
// Pins the JS district-id helpers to the shared R/JS fixtures
// (src/loaders/R/common/districts_fixtures.csv). The R twin is
// scripts/test-districts.R — run BOTH after touching either implementation.

import fs from "node:fs";
import { csvParse } from "d3-dsv";
import { councilSelfgovIdFromMajorId, selfgovIdFromRawDistrictId } from "../src/components/election-utils.js";

const fixtures = csvParse(fs.readFileSync("src/loaders/R/common/districts_fixtures.csv", "utf8"))
  // "dotted" pins the R-only ingest helper council_maj_id_from_dotted();
  // there is no JS twin (raw codes never reach the client).
  .filter(f => f.fn !== "dotted");
let failed = 0;
for (const { fn, input, expected, note } of fixtures) {
  const got = fn === "maj" ? councilSelfgovIdFromMajorId(input) : selfgovIdFromRawDistrictId(input);
  if (got !== expected) {
    failed++;
    console.error(`✗ ${fn}(${input}) = ${got}, expected ${expected}  (${note})`);
  }
}
if (failed) {
  console.error(`\n${failed}/${fixtures.length} fixture(s) FAILED`);
  process.exit(1);
}
console.log(`All ${fixtures.length} district-id fixtures passed (JS).`);
